### DISCUSSION TITLE
Discriminated Union in Schema?

### DISCUSSION THREAD

**Discussion by princemuel** [colinhacks/zod#2761] [Q&A] [answered]
Hi there. Please I have a question. I have a checkout form where there are two forms of payment, inCash and eMoney. When eMoney is selected, I want to show two input fields: card num and pin, but if inCash is selected, nothing should be shown.

How do I setup the schema to have something like a discriminated union (I guess) so that card number and card pin should be required if the payment prop in the schema === eMoney.

```js
// payment === eMoney
{ payment: 'eMoney', cardPin: z.number(), cardNumber: z.number(),}
// payment === cash
{ payment: 'cash' }
```

PS: I'm a newbie to Zod, and I feel like I'm not asking the question right. I'd appreciate any suggestion on how to improve it. Thanks 🙏

**Comment by JacobWeisenburger** [comment]
Give this a try:
https://github.com/colinhacks/zod#discriminated-unions

**Reply by princemuel** [reply]
Yeah, I've read the docs, but I'm not really sure how to implement that. Some methods I've seen call for refine or super refine?

**Comment by JacobWeisenburger** [comment, marked as answer]
Please let me know if there is something I am missing.

```ts
const myUnion = z.discriminatedUnion( 'payment', [
    z.object( {
        payment: z.literal( 'cash' ),
    } ),
    z.object( {
        payment: z.literal( 'eMoney' ),
        cardPin: z.number(),
        cardNumber: z.number(),
    } ),
] )

console.log(
    myUnion.parse( { payment: 'cash' } )
) // { payment: 'cash' }

console.log(
    myUnion.parse( {
        payment: 'eMoney',
        cardPin: 1234,
        cardNumber: 1234567890,
    } )
) // { payment: 'eMoney', cardPin: 1234, cardNumber: 1234567890 }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
